[ao][sparsity] Base class for Data Sparsifier

macandro96 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79535
* #79534
* #79254
* #79253
* #79427
* #79252
* __->__ #79251

Base Data Sparsifier class for all Data sparsifiers.
The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
to prepare for sparsification.
In this case, mask (and parametrizations) is owned by the class and not by the user.
Specifically, the container object inside the class maintains the mask and parametrizations of the input data

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

Differential Revision: [D37164120](https://our.internmc.facebook.com/intern/diff/D37164120)

Commit:
[ao][sparsity] Base class for Data Sparsifier
Base Data Sparsifier class for all Data sparsifiers.
The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
to prepare for sparsification.
In this case, mask (and parametrizations) is owned by the class and not by the user.
Specifically, the container object inside the class maintains the mask and parametrizations of the input data

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

HDCharles reviewed (commented):
lgtm

albanD commented on test/allowlist_for_publicAPI.json:
```diff
@@ -369,6 +369,14 @@
     "Any",
     "List"
   ],
+  "torch.ao.sparsity.experimental.data_sparsifier.base_data_sparsifier": [
```
No new entry should be added to this file.
In this case, you most likely just want to define `__all__` properly for this submodule.

macandro96 replied:
These are generic typing packages [List, Any] that are not actually exposed by the submodule (but used by them). Should this still be inside __all__?

albanD replied:
I meant that you should have an `__all__` that contain all the public APIs you want to expose and these should not be there.
if you do that, you won't have to add anything to this allowlist.

albanD replied:
Things like `Any` or `List` are not part of the things your module is exposing publicly (I hope) so these should not be in the `__all__`.

z-a-f replied:
> 

I think the default behavior in Python is that any imports within a package would be exposed globally to the user. Please, correct me. Here is a basic example that I have, in case it helps understand what I mean: https://colab.research.google.com/drive/1q-p8zfv_p7Fg1AxR4gWveQszNm-ENCZu?usp=sharing

albanD replied:
Yes everything is exposed. but there is nothing we can do about that (unless we force everyone to `del Any, List` etc at the end of every single python file). That's why our public API defition is based on `__all__` when it exists.
Note that when you do `from foo import *`, the `__all__` is what will control what gets imported. So in that sense it does control what gets exposed to the user.

macandro96 replied:
added __all__ into the .py file

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -0,0 +1,76 @@
+    def step(self):
+        pass
+
+    def update_mask(self, name, data, **kwargs):
```
Shouldn't that be an `abstractmethod`?

macandro96 replied:
Yep - will add that.

macandro96 replied:
added

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -0,0 +1,76 @@
+    def get_data(self, name: str):
+        pass
+
+    def state_dict(self):
+        pass
+
+    def load_state_dict(self, state_dict):
+        pass
+
+    def __setstate__(self, state):
+        pass
+
+    def __getstate__(self):
+        pass
```
If these are the placeholders, we can just remove them from this PR, and add them at a later diff

macandro96 replied:
removed them and added them in a later diff

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -0,0 +1,76 @@
+        """
+        pass
+
+    def get_data(self, name: str):
+        pass
```
Can we add a docstring here. Even if there is no implementation as of now, at least we would get an idea why this method is needed

macandro96 replied:
done.

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -0,0 +1,76 @@
+    def __getstate__(self):
+        pass
+
+    def __repr__(self):
+        pass
+
+    def get_mask(self, name: str):
+        pass
+
+    def squash_mask(self, *args, **kwargs):
+        pass
+
+    def step(self):
+        pass
```
Same "docstring" comment as in the `get_data`

macandro96 replied:
done

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -0,0 +1,76 @@
+    to prepare for sparsification.
+    In this case, mask (and parametrizations) is owned by the class and not by the user.
+    Specifically, the container object inside the class maintains the mask and parametrizations of the input data
+    Args:
```
Space before `Args:`. See [documentation here](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html?highlight=Args#type-annotations) (refer to the "google style")

macandro96 replied:
Noted

jerryzh168 replied:
example: https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/conv.py#L671

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -0,0 +1,76 @@
+    In this case, mask (and parametrizations) is owned by the class and not by the user.
+    Specifically, the container object inside the class maintains the mask and parametrizations of the input data
+    Args:
+        - data_list [List[Tuple[str, Any]]]: list of (name, data) tuples to sparsify.
+        Currently - operates on embeddings and torch tensors, parameters
+        Internally, a container module handles the data sparsification.
+
+        - defaults [dict]: default configurations will be attached to the
+            configuration. Only the keys that don't exist in the `config` will
+            be updated.
```
1. Don't add `-` in front of the parameter name
2. The data type should be in regular parenthesis `(` and `)`
3. The data type should be human readable, not necessarily `typing` instance. For example, use `list of int's` instead of `List[int]`
4. The continuation in the line should be indented.

One option to rewrite this:

```
    Args:
        data_list (list of tuples):
            list of (name, data) tuples to sparsify.
            Currently - operates on embeddings and torch tensors, parameters
            Internally, a container module handles the data sparsification.
        defaults [dict]:
            default configurations will be attached to the
            configuration. Only the keys that don't exist in the `config` will
            be updated.
```

macandro96 replied:
Noted :)

macandro96 replied:
modified

Commit:
Update on "[ao][sparsity] Base class for Data Sparsifier"


Base Data Sparsifier class for all Data sparsifiers.
The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
to prepare for sparsification.
In this case, mask (and parametrizations) is owned by the class and not by the user.
Specifically, the container object inside the class maintains the mask and parametrizations of the input data

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Base class for Data Sparsifier"


Base Data Sparsifier class for all Data sparsifiers.
The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
to prepare for sparsification.
In this case, mask (and parametrizations) is owned by the class and not by the user.
Specifically, the container object inside the class maintains the mask and parametrizations of the input data

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

jerryzh168 reviewed (commented):
LGTM, but I'll let others to accept since I'm not sure what method we should have in the sparsifier

HDCharles reviewed (approved):
lgtm

Commit:
Update on "[ao][sparsity] Base class for Data Sparsifier"


Base Data Sparsifier class for all Data sparsifiers.
The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
to prepare for sparsification.
In this case, mask (and parametrizations) is owned by the class and not by the user.
Specifically, the container object inside the class maintains the mask and parametrizations of the input data

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Base class for Data Sparsifier"


Base Data Sparsifier class for all Data sparsifiers.
The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
to prepare for sparsification.
In this case, mask (and parametrizations) is owned by the class and not by the user.
Specifically, the container object inside the class maintains the mask and parametrizations of the input data

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

z-a-f reviewed (approved):
lgtm

Commit:
Update on "[ao][sparsity] Base class for Data Sparsifier"


Base Data Sparsifier class for all Data sparsifiers.
The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
to prepare for sparsification.
In this case, mask (and parametrizations) is owned by the class and not by the user.
Specifically, the container object inside the class maintains the mask and parametrizations of the input data

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

Note::
    Workflow failure *linux-docs / build-docs (python)* - This diff creates a new file in the
    experimental folder which does not have docs associated with it.
    This can be fixed by making the experimental folder (in torch/ao/sparsity/)
    private which will be done so later in a separate PR.

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Base class for Data Sparsifier"


Base Data Sparsifier class for all Data sparsifiers.
The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
to prepare for sparsification.
In this case, mask (and parametrizations) is owned by the class and not by the user.
Specifically, the container object inside the class maintains the mask and parametrizations of the input data

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

Note::
    Workflow failure *linux-docs / build-docs (python)* - This diff creates a new file in the
    experimental folder which does not have docs associated with it.
    This can be fixed by making the experimental folder (in torch/ao/sparsity/)
    private which will be done so later in a separate PR.

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Base class for Data Sparsifier"


Base Data Sparsifier class for all Data sparsifiers.
The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
to prepare for sparsification.
In this case, mask (and parametrizations) is owned by the class and not by the user.
Specifically, the container object inside the class maintains the mask and parametrizations of the input data

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

Note::
    Workflow failure *linux-docs / build-docs (python)* - This diff creates a new file in the
    experimental folder which does not have docs associated with it.
    This can be fixed by making the experimental folder (in torch/ao/sparsity/)
    private which will be done so later in a separate PR.

[ghstack-poisoned]

macandro96 commented:
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164120).

Commit:
Update on "[ao][sparsity] Base class for Data Sparsifier"


Base Data Sparsifier class for all Data sparsifiers.
The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
to prepare for sparsification.
In this case, mask (and parametrizations) is owned by the class and not by the user.
Specifically, the container object inside the class maintains the mask and parametrizations of the input data

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

Note::
    Workflow failure *linux-docs / build-docs (python)* - This diff creates a new file in the
    experimental folder which does not have docs associated with it.
    This can be fixed by making the experimental folder (in torch/ao/sparsity/)
    private which will be done so later in a separate PR.

Differential Revision: [D37164120](https://our.internmc.facebook.com/intern/diff/D37164120)

[ghstack-poisoned]

macandro96 commented:
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164120).

Commit:
Update on "[ao][sparsity] Base class for Data Sparsifier"


Base Data Sparsifier class for all Data sparsifiers.
The abstract class accepts raw torch tensors / embedding / embedding bags (refer to SUPPORTED_TYPES above)
to prepare for sparsification.
In this case, mask (and parametrizations) is owned by the class and not by the user.
Specifically, the container object inside the class maintains the mask and parametrizations of the input data

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

Differential Revision: [D37164120](https://our.internmc.facebook.com/intern/diff/D37164120)

[ghstack-poisoned]

macandro96 commented:
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164120).
